diskplorer.py

```diff
@@ -205,7 +215,29 @@ def global_section():
             job_names = generate_job_names(f'job(r_idx={read_iops_step},w_idx={write_bw_step},write_bw={write_bw},r_iops={read_iops})')
             read_iops = max(read_iops, 10)   # no point in a write-only test
             nr_cpus = args.cpus
+            read_offset_str = ''
+
             if write_bw > 0:
+                if args.run_sequential_trim:
+                    out(textwrap.dedent(f'''\
+                        [prepare_data_for_trim(r_idx={read_iops_step},w_idx={write_bw_step},write_bw={write_bw},r_iops={read_iops})]
+                        '''))
+                    out(group_introducer)
+                    out(textwrap.dedent(f'''\
+                        readwrite=write
+                        runtime={args.trim_offset_time_seconds}s
+                        blocksize={args.write_buffer_size}
+                        iodepth={args.write_concurrency}
+                        rate={write_bw}
+                        '''))
+                    read_offset = write_bw*(args.test_step_time_seconds+args.trim_offset_time_seconds)
+                    read_offset_str = f'offset={align_up(read_offset, args.read_buffer_size)}'
+                    write_offset = write_bw*args.trim_offset_time_seconds
+                    write_offset_str = f'offset={align_up(write_offset, args.write_buffer_size)}'
```

Comment:
Is this additional write stream part of the normal test? If so it affects it.

I don't understand the runtime part.

Reply:
Firstly, I will describe what was my intention. If the user specified `args.sequential_trim=true`, then:
1. Before ordinary read and write jobs are run for `(write_bw_step, read_iops_step)`, then a new job called `prepare_data_for_trim` is executed. It writes the data according to the following requirement:
> Because we can't trim into the past, we'll need to add a k-second write/read job (that isn't measured) followed by the regular 30-second write/read/trim-behind job. The offset= fio command can be used to orchestrate the second write job wrt the first.
2. The new job uses `out(group_introducer) == stonewall + new_group` to ensure that all previous jobs finished before it started. From FIO docs - [stonewall](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-stonewall): 
> Wait for preceding jobs in the job file to exit, before starting this one. Can be used to insert serialization points in the job file. A stone wall also implies starting a new reporting group, see [group_reporting](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-group_reporting).
4. The first ordinary write or read job also contains `stonewall + new_group` to ensure that `prepare_data_for_trim` finished its work before read+write workloads started.


Regarding runtime and offset parameters:
- `runtime={args.trim_offset_time_seconds}s` is used in case of `prepare_data_for_trim` to write the data that will be trimmed by trim workload in k seconds
- `write_offset = write_bw*args.trim_offset_time_seconds` instructs the write workload to start writing data right after the data that was written by `prepare_data_for_trim`
- `read_offset = write_bw*(args.test_step_time_seconds+args.trim_offset_time_seconds)` instructs read workload to not touch any of the data that will be trimmed. I thought that trim workload will discard `(write_bw*args.test_step_time_seconds)` of bytes. However, I still added the preparation time - not sure why.


Regarding affecting the test case by `prepare_data_for_trim` - I thought that waiting until it finishes before running read+write workloads and usage of different reporting groups is sufficient. Also, latency-postprocess.py contains:
```
if name == 'prepare' or name.startswith('prepare_data_for_trim') or name.startswith('trim_data'):
            continue
```

There is a chance, that I misunderstood how FIO groups the measurements.

Reply:
> Firstly, I will describe what was my intention. If the user specified `args.sequential_trim=true`, then:
> 
> 1. Before ordinary read and write jobs are run for `(write_bw_step, read_iops_step)`, then a new job called `prepare_data_for_trim` is executed. It writes the data according to the following requirement:
> 
> > Because we can't trim into the past, we'll need to add a k-second write/read job (that isn't measured) followed by the regular 30-second write/read/trim-behind job. The offset= fio command can be used to orchestrate the second write job wrt the first.


Ok, makes sense. But please add comments explaining it.

Maybe it should be bounded by write amount, not runtime. It makes sense to use the same write rate to not disrupt the flow.


> 
> 2. The new job uses `out(group_introducer) == stonewall + new_group` to ensure that all previous jobs finished before it started. From FIO docs - [stonewall](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-stonewall):
> 
> > Wait for preceding jobs in the job file to exit, before starting this one. Can be used to insert serialization points in the job file. A stone wall also implies starting a new reporting group, see [group_reporting](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-group_reporting).
> 

I think it's correct, but best to observe using iostat -x (which shows discards) and with blktrace (only useful for really short periods). It's important to have trust in the measurement tools.


> 4. The first ordinary write or read job also contains `stonewall + new_group` to ensure that `prepare_data_for_trim` finished its work before read+write workloads started.
> 
> Regarding runtime and offset parameters:
> 
> * `runtime={args.trim_offset_time_seconds}s` is used in case of `prepare_data_for_trim` to write the data that will be trimmed by trim workload in k seconds
> * `write_offset = write_bw*args.trim_offset_time_seconds` instructs the write workload to start writing data right after the data that was written by `prepare_data_for_trim`

But it assumes the bw*time = what you wanted. It's probably okay but can be subject to rounding. Better to limit it by data size.


> * `read_offset = write_bw*(args.test_step_time_seconds+args.trim_offset_time_seconds)` instructs read workload to not touch any of the data that will be trimmed.

Ok. What about the end of the range? Is it shifted into not-written territory, or does it stay?

I forgot how diskplorer ensures reads access previously-written addresses.

> I thought that trim workload will discard `(write_bw*args.test_step_time_seconds)` of bytes. However, I still added the preparation time - not sure why.
> 
> Regarding affecting the test case by `prepare_data_for_trim` - I thought that waiting until it finishes before running read+write workloads and usage of different reporting groups is sufficient. Also, latency-postprocess.py contains:

I meant something else - that there's a separate write stream during the measured time period. If there isn't we're good.


> 
> ```
> if name == 'prepare' or name.startswith('prepare_data_for_trim') or name.startswith('trim_data'):
>             continue
> ```
> 
> There is a chance, that I misunderstood how FIO groups the measurements.

It's tricky and therefore important to observe runs and see that what happens matches expectations. That's what I did when writing the thing.

Reply:
> Ok, makes sense. But please add comments explaining it.
> 
> Maybe it should be bounded by write amount, not runtime. It makes sense to use the same write rate to not disrupt the flow.

Sure. I added a comment with explanations to the code. Moreover, I used the following syntax to ensure, that the desired amount of data is written by the job - `pre_written_data_size` is `write_bw*k_seconds` aligned according to the requirements.
```
time_based=0
runtime=0
size={pre_written_data_size}
```


> Ok. What about the end of the range? Is it shifted into not-written territory, or does it stay?
> 
> I forgot how diskplorer ensures reads access previously-written addresses.

The end of the range is not changed according to FIO documentation that describes [offset](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-offset).

> Start I/O at the provided offset in the file, given as either a fixed size in bytes, zones or a percentage. [...] Data before the given offset will not be touched. This effectively caps the file size at real_size - offset. Can be combined with [size](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-size) to constrain the start and end range of the I/O workload.


I assumed that either `--prefill` or `--size-limit` is used by diskplorer to ensure that read operations access previously written addresses.

With `--prefill` flag diskplorer writes the whole disk as soon as it is invoked. This way we can use `rw=randread` and don't care about the place where randread is done.
`help='Prefill entire disk, defeats incorrect results due to discard (default)'`

With `--size-limit` I would assume, that the data was written before and we constrain diskplorer to use only that part of the storage.

The prepare job is defined as follows:
```
[prepare]
readwrite=write
time_based=0
blocksize=2MB
iodepth=4
runtime=0
```

The read job is defined as:
```
[read_job]
readwrite=randread
blocksize={args.read_buffer_size}
iodepth={args.read_concurrency}
rate_iops={this_cpu_read_iops}
rate_process=poisson
{read_offset_str}
```

Reply:
> > Ok, makes sense. But please add comments explaining it.
> > Maybe it should be bounded by write amount, not runtime. It makes sense to use the same write rate to not disrupt the flow.
> 
> Sure. I added a comment with explanations to the code. Moreover, I used the following syntax to ensure, that the desired amount of data is written by the job - `pre_written_data_size` is `write_bw*k_seconds` aligned according to the requirements.
> 
> ```
> time_based=0
> runtime=0
> size={pre_written_data_size}
> ```
> 
> > Ok. What about the end of the range? Is it shifted into not-written territory, or does it stay?
> > I forgot how diskplorer ensures reads access previously-written addresses.
> 


> The end of the range is not changed according to FIO documentation that describes [offset](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-offset).
> 


Ack

> > Start I/O at the provided offset in the file, given as either a fixed size in bytes, zones or a percentage. [...] Data before the given offset will not be touched. This effectively caps the file size at real_size - offset. Can be combined with [size](https://fio.readthedocs.io/en/latest/fio_doc.html#cmdoption-arg-size) to constrain the start and end range of the I/O workload.
> 
> I assumed that either `--prefill` or `--size-limit` is used by diskplorer to ensure that read operations access previously written addresses.
> 

Ah, I forgot about prefill. So it's enough to exclude anything that may have been discarded.

> With `--prefill` flag diskplorer writes the whole disk as soon as it is invoked. This way we can use `rw=randread` and don't care about the place where randread is done. `help='Prefill entire disk, defeats incorrect results due to discard (default)'`
> 
> With `--size-limit` I would assume, that the data was written before and we constrain diskplorer to use only that part of the storage.

Yes, the main concern is that the default workload is sensible.


> 
> The prepare job is defined as follows:
> 
> ```
> [prepare]
> readwrite=write
> time_based=0
> blocksize=2MB
> iodepth=4
> runtime=0
> ```
> 
> The read job is defined as:
> 
> ```
> [read_job]
> readwrite=randread
> blocksize={args.read_buffer_size}
> iodepth={args.read_concurrency}
> rate_iops={this_cpu_read_iops}
> rate_process=poisson
> {read_offset_str}
> ```

Looks good.